[Quant][fx] Add default configs for fixed qparams ops

andrewor14 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #80184

Summary: This commit adds qconfigs with special observers for fixed
qparams ops in get_default_qconfig_mapping and
get_default_qat_qconfig_mapping. For correctness, we also require
users to use these special observers if we detect these fixed
qparams ops in prepare.

BC-breaking notes:

Before (fails after this PR):
```
from torch.ao.quantization.quantize_fx import prepare_fx

model = ModelWithFixedQParamsOps()
qconfig_mapping = QConfigMapping()
example_inputs = ...
prepare_fx(model, qconfig_mapping, example_inputs)
```

After:
```
from torch.ao.quantization import get_default_qconfig_mapping
from torch.ao.quantization.quantize_fx import prepare_fx

model = ModelWithFixedQParamsOps()
qconfig_mapping = get_default_qconfig_mapping()
example_inputs = ...
prepare_fx(model, qconfig_mapping, example_inputs)
```

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

Differential Revision: [D37396379](https://our.internmc.facebook.com/intern/diff/D37396379)

andrewor14 committed:
[Quant][fx] Add default configs for fixed qparams ops

Summary: This commit adds qconfigs with special observers for fixed
qparams ops in get_default_qconfig_mapping and
get_default_qat_qconfig_mapping. For correctness, we also require
users to use these special observers if we detect these fixed
qparams ops in prepare.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Add default configs for fixed qparams ops"


Summary: This commit adds qconfigs with special observers for fixed
qparams ops in get_default_qconfig_mapping and
get_default_qat_qconfig_mapping. For correctness, we also require
users to use these special observers if we detect these fixed
qparams ops in prepare.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Add default configs for fixed qparams ops"


Summary: This commit adds qconfigs with special observers for fixed
qparams ops in get_default_qconfig_mapping and
get_default_qat_qconfig_mapping. For correctness, we also require
users to use these special observers if we detect these fixed
qparams ops in prepare.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37396379).

andrewor14 committed:
Update on "[Quant][fx] Add default configs for fixed qparams ops"


Summary: This commit adds qconfigs with special observers for fixed
qparams ops in get_default_qconfig_mapping and
get_default_qat_qconfig_mapping. For correctness, we also require
users to use these special observers if we detect these fixed
qparams ops in prepare.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

Differential Revision: [D37396379](https://our.internmc.facebook.com/intern/diff/D37396379)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37396379).

jerryzh168 commented on torch/ao/quantization/observer.py:
```diff
@@ -1512,3 +1546,19 @@ def load_observer_state_dict(mod, obs_dict):
 the operator
 """
+
+_FIXED_QPARAMS_OP_TO_OBSERVER: Dict[Union[Callable, str], _PartialWrapper] = {
```
I feel this should probably live in somewhere else, it's a bit weird to put this mapping in observer.py, maybe move this to qconfig_mapping.py?

jerryzh168 commented on torch/ao/quantization/qconfig_mapping.py:
```diff
@@ -73,13 +74,24 @@ def _get_default_qconfig_mapping(is_qat: bool, backend: str, version: int):
 
-def get_default_qconfig_mapping(backend="fbgemm", version=0):
+    # Use special observers for ops with fixed qparams
+    fixed_qparams_observer_to_qconfig: Dict[Any, QConfigAny] = {}
```
what is this? I didn't see the code that modifies this dictionary

andrewor14 replied:
Oh I forgot to populate it in the for loop below

jerryzh168 commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -1316,6 +1322,23 @@ def insert_observers_for_model(
+    }
+    for node in model.graph.nodes:
+        if node.target in _FIXED_QPARAMS_OP_TO_OBSERVER:
+            qconfig = get_object_type_qconfig(qconfig_mapping, node.target, qconfig_mapping.global_qconfig)
```
I feel we should get the qconfig for this node from the generated `qconfig_map`, since user may use different ways of configuration to set the qconfig for these nodes

jerryzh168 replied:
in L1517 of this file:
```
qconfig_map = generate_qconfig_map(model, modules, model.graph, qconfig_mapping, node_name_to_scope)
```
we should be using this one I think, also maybe makes sense to rename it since now we have `qconfig_mapping`

andrewor14 replied:
I see, OK. I feel we really need to rename it now :)

andrewor14 replied:
I see, OK. Yes I feel we really need to rename it now :)

jerryzh168 commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -1435,6 +1458,8 @@ def prepare(
     assert(isinstance(qconfig_mapping, QConfigMapping))
     assert(isinstance(equalization_config, QConfigMapping))
+
+    _validate_fixed_qparams_qconfigs(model, qconfig_mapping)
```
here is where we verify that qconfig is supported by backend_config: https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fx/prepare.py#L1199, I think we can put the validation close to this line so that it's easier to modify the logic in the future

jerryzh168 reviewed (approved):
thanks, looks good overall, had a few comments inline. please also make sure to do ghimport and fix the internal callsites as well

andrewor14 committed:
Update on "[Quant][fx] Add default configs for fixed qparams ops"


Summary: This commit adds qconfigs with special observers for fixed
qparams ops in get_default_qconfig_mapping and
get_default_qat_qconfig_mapping. For correctness, we also require
users to use these special observers if we detect these fixed
qparams ops in prepare.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

Differential Revision: [D37396379](https://our.internmc.facebook.com/intern/diff/D37396379)

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Add default configs for fixed qparams ops"


Summary: This commit adds qconfigs with special observers for fixed
qparams ops in get_default_qconfig_mapping and
get_default_qat_qconfig_mapping. For correctness, we also require
users to use these special observers if we detect these fixed
qparams ops in prepare.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

Differential Revision: [D37396379](https://our.internmc.facebook.com/intern/diff/D37396379)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37396379).

andrewor14 committed:
Update on "[Quant][fx] Add default configs for fixed qparams ops"


Summary: This commit adds qconfigs with special observers for fixed
qparams ops in get_default_qconfig_mapping and
get_default_qat_qconfig_mapping. For correctness, we also require
users to use these special observers if we detect these fixed
qparams ops in prepare.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168, vkuzo

Subscribers: jerryzh168, vkuzo

Differential Revision: [D37396379](https://our.internmc.facebook.com/intern/diff/D37396379)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37396379).
